environment/environment.go

```diff
@@ -0,0 +1,71 @@
+package environment
+
+import (
+	"fmt"
+
+	"github.com/sourcegraph/conc"
+)
+
+type deferFn struct {
+	msg string
+	fn  func() error
+}
+
+// Env controls goroutines and deferred functions. Think of it as the "main thread."
+// It is not goroutine-safe.
+type Env struct {
+	deferFns []*deferFn
+	wg       *conc.WaitGroup
+}
+
+func New() *Env {
+	return &Env{
+		deferFns: []*deferFn{},
+		wg:       conc.NewWaitGroup(),
+	}
+}
+
+// Go runs fn in a separate goroutine. Close will block until fn returns.
+func (s *Env) Go(fn func()) {
+	s.wg.Go(fn)
+}
+
+// Defer saves fn, which will be run on Close.
```

Comment:
Something like this recovers a more regular resource freeing schedule:

```go
// GoThen runs fn, and then deferred, in a separate goroutine.
//
// Close will block until both functions return.
func (s *Env) GoThen(fn func(), deferred func()) {
	s.wg.Go(func() {
		defer deferred()
		fn()
	})
}

// GoThenErr runs fn, and then deferred, in a separate goroutine.
//
// If deferred returns an error, it will be collected on the error channel.
//
// Close will block until both functions return.
func (s *Env) GoThenErr(fn func(), deferred func() error, errMsg string) {
	s.GoThen(fn, func() {
		err := deferred()
		if err != nil {
			s.errCh <- fmt.Errorf("%s: %v", errMsg, err)
		}
	})
}
```

Reply:
Insightful questions.

> Are we concerned about resource accumulation with this scheme?

No, this is actually a little more memory efficient than before. Previously, we were spawning a bunch of goroutines, blocking in each goroutine on `<-ctx.Done()`, and then `defer`-ing the cleanups. Now, we don't need to spawn unnecessary goroutines: we just pass around the `Env` object and add `defer`s to it.

Side note: this means that cleanup for some resources won't run in parallel anymore, but if that becomes an issue, we can easily compose `Env` objects.

> Possible future deadlock (eg, multiple attempts to file handles on a db file could fail)?

I don't think this can happen if the `Env` is used properly. Resources created first are cleaned up last.

> Something like this recovers a more regular resource freeing schedule

I agree with your general points about how defer-ed clean-ups should live close to the resource initialization. I think the `Env` preserves this pattern in practice (`env.Defer` often directly follows `env.Go`, for example).